torch/csrc/jit/runtime/interpreter.cpp

```diff
@@ -1245,9 +1274,18 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
               push(stack, true);
             } else {
               auto t = stack.back().toTensor();
-              const TypePtr& expected = af.types[inst.X];
               auto pttp = tensorTypeInCurrentExecutionContext(t);
-              push(stack, pttp->isSubtypeOf(expected));
+              const TypePtr& expected = af.types[inst.X];
+              auto expected_type = expected->cast<TensorType>();
+              bool pass = true;
+              if (t.defined()) {
+                pass &= bindSymbolicShapes(
+                    af, t.sizes(), expected_type->symbolic_sizes());
+                if (pass) {
+                  pttp = expected_type->merge(pttp, false);
```

Comment:
I think the composition of bindSymbolicShapes, merging and isSubtypeOf here is not ideal. First, it is a bit hard to follow direct intent of tracking whether the guard matches. I am still not sure what merge, and `false` is suppose to be here. Second, this is in performance critical code. `isSubtypeOf` checks are slow, and there is a lot we can do here to combine the bindSymbolicShape, and isSubtypeOf together to make a single pass over the tensor. Finally, I am confused as to whether returning false from bindSymbolicShapes will cause the guard to fail. I think it should, since we found evidence of the program no longer passing its profiled model, but it doesn't directly lead to a push `false`, so I am not sure.

Reply:
>  it doesn't directly lead to a push false, so I am not sure.

yes it's not too obvious that `expected_type` either contains static or symbolic symbols and `isSubtypeOf` will fail when comparing with `pttp` which only has static symbols unless they are either the same or we run `bindSymbolicShapes` . I explicitly added `&& bound_successfully &&` to push.